File: php/NemesisShield.php
<?php
// Nemesis Shield - Sentinel SDK for PHP (native: local shape + cached policy + inline blocking).
// Learns your app's normal behavior; in enforce mode blocks off-baseline requests (auth bypass, path
// traversal, scanners, unusual methods) before your app runs. PHP is stateless per-request, so the
// compiled policy is cached to a temp file with a short TTL and refreshed on demand. Fail-open.
//
// Raw:      if (NemesisShield::guard($token)) return;  ... your app ...  NemesisShield::observe($token);
// Laravel/Symfony/PSR-15: see NemesisShieldMiddleware below.

class NemesisShield
{
    const ENDPOINT = 'https://shield.nemesislabs.xyz/api/v1/sketches';
    const TTL = 2; // seconds; refresh the cached policy at most this often

    /**
     * The sketches endpoint. Defaults to Nemesis Shield cloud; override with the
     * NEMESIS_ENDPOINT env var or a NEMESIS_SHIELD_ENDPOINT constant to point at a
     * self-hosted / on-prem Shield (or a local mock in tests). Backward-compatible:
     * with neither set, the cloud const is used exactly as before.
     */
    private static function endpoint(): string
    {
        if (defined('NEMESIS_SHIELD_ENDPOINT') && NEMESIS_SHIELD_ENDPOINT) return NEMESIS_SHIELD_ENDPOINT;
        $env = getenv('NEMESIS_ENDPOINT');
        return ($env !== false && $env !== '') ? $env : self::ENDPOINT;
    }

    // Safe-unlock (break-glass): paths never blocked, so a still-learning baseline can't lock
    // operators out of the doors they need to fix it. Prefix-matched on the pathname. Override/extend
    // with the NEMESIS_SHIELD_BOOTSTRAP env (comma-separated).
    const DEFAULT_BOOTSTRAP = ['/login', '/signin', '/sign-in', '/auth', '/oauth', '/session', '/wp-login.php', '/wp-admin'];

    private static function bootstrap(): array
    {
        $env = getenv('NEMESIS_SHIELD_BOOTSTRAP');
        if ($env !== false && trim($env) !== '') {
            return array_values(array_filter(array_map('trim', explode(',', $env))));
        }
        return self::DEFAULT_BOOTSTRAP;
    }

    /** Break-glass: is this path on the never-block bootstrap allow-list? Prefix match. */
    public static function neverBlock(string $path): bool
    {
        $p = strtolower(explode('#', explode('?', $path, 2)[0], 2)[0]);
        foreach (self::bootstrap() as $b) {
            if ($b !== '' && strpos($p, strtolower($b)) === 0) return true;
        }
        return false;
    }

    private static function cacheFile(string $token): string
    {
        return sys_get_temp_dir() . '/nemesis_' . substr(sha1($token), 0, 16) . '.json';
    }

    public static function normalizePath(string $path): string
    {
        $path = explode('?', $path, 2)[0];
        $path = explode('#', $path, 2)[0];
        $segs = explode('/', $path);
        foreach ($segs as $i => $s) {
            if ($s === '') continue;
            if (strpos($s, '..') !== false) { $segs[$i] = '{traversal}'; continue; } // keeps ".." out of telemetry
            switch (self::kindOf($s)) {
                case 'int': case 'float': $segs[$i] = '{int}'; break;
                case 'uuid': $segs[$i] = '{uuid}'; break;
                case 'hex': $segs[$i] = '{hex}'; break;
                case 'base64': $segs[$i] = '{token}'; break;
                case 'alnum': if (strlen($s) >= 12 && !preg_match('/[A-Za-z]{7,}/', $s)) $segs[$i] = '{id}'; break;
            }
        }
        $out = implode('/', $segs);
        return $out === '' ? '/' : $out;
    }

    // Canonical value taxonomy - must match the shared engine (tokenize.ts) byte-for-byte.
    private static function kindOf($v): string
    {
        if ($v === null || $v === '') return 'empty';
        if (is_bool($v)) return 'bool';
        $s = (string)$v;
        if ($s === '') return 'empty';
        if (preg_match('/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i', $s)) return 'uuid';
        if (preg_match('/^[^@\s]+@[^@\s]+\.[^@\s]+$/', $s)) return 'email';
        if (preg_match('#^https?://\S+$#i', $s)) return 'url';
        if (preg_match('/^(\d{1,3}\.){3}\d{1,3}$/', $s)) return 'ipv4';
        if (preg_match('/^[0-9a-f:]+:[0-9a-f:]+$/i', $s)) return 'ipv6';
        if (preg_match('/^[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+$/', $s)) return 'jwt';
        if (preg_match('/^\d{4}-\d{2}-\d{2}([T ]\d{2}:\d{2}(:\d{2})?)?/', $s)) return 'date';
        if (preg_match('/^-?\d+$/', $s)) return 'int';
        if (preg_match('/^-?\d*\.\d+$/', $s)) return 'float';
        if (strlen($s) >= 16 && preg_match('/^[0-9a-f]+$/i', $s)) return 'hex';
        if (strlen($s) >= 16 && preg_match('#^[A-Za-z0-9+/]+={0,2}$#', $s) && !preg_match('/[A-Za-z]{7,}/', $s)) return 'base64';
        if (preg_match('/^[A-Za-z]+$/', $s)) return 'alpha';
        if (preg_match('/^[A-Za-z0-9]+$/', $s)) return 'alnum';
        return 'string';
    }

    private static function fnv1a(string $s): string
    {
        $h = 0x811c9dc5;
        for ($i = 0, $n = strlen($s); $i < $n; $i++) {
            $h ^= ord($s[$i]);
            $h = ($h * 0x01000193) & 0xffffffff;
        }
        return str_pad(dechex($h), 8, '0', STR_PAD_LEFT);
    }

    // Analytics / click-tracking query params carry no application logic and are the main cause of
    // shape explosion (every ad/campaign link adds different ones). Stripped from the signature so a
    // UTM'd request matches the bare route, while real params (and any attack in them) stay modeled.
    // Shared, identical across every Nemesis Shield SDK so the shape hash matches everywhere.
    private static function isTrackingParam($name): bool
    {
        $n = strtolower((string) $name);
        foreach (array('utm_', 'mtm_', 'pk_', 'hsa_', 'matomo_', 'piwik_', 'ga_') as $p) {
            if (strpos($n, $p) === 0) return true;
        }
        return in_array($n, array('gclid', 'gbraid', 'wbraid', 'dclid', 'gclsrc', 'fbclid', 'msclkid', 'twclid', 'ttclid', 'yclid', 'igshid', 'scid', 'wickedid', '_ga', '_gl', '_hsenc', '_hsmi', 'mc_cid', 'mc_eid', 'vero_id', 'vero_conv', 'oly_anon_id', 'oly_enc_id', '_openstat', 'rb_clickid', 's_cid', 'epik', 'sccid'), true);
    }

    public static function buildSketch(string $method, string $path, array $query, bool $authed, int $status): array
    {
        $route = self::normalizePath($path);
        $names = array_filter(array_keys($query), function ($k) { return !self::isTrackingParam($k); });
        sort($names);
        $params = [];
        foreach ($names as $k) {
            $nested = is_array($query[$k]);
            $v = $nested ? ($query[$k][0] ?? '') : $query[$k];
            $params[] = ['name' => (string)$k, 'kind' => self::kindOf($v), 'nested' => $nested];
        }
        // canonical shape input: keys sorted (auth, method, params, route); params [name, kind, nested];
        // status excluded so enforcement decides BEFORE the response exists.
        // plain closure (not fn=>) so the SDK runs on PHP 7.2+ - matching Laravel 6's own floor.
        $canonParams = array_map(function ($p) { return [$p['name'], $p['kind'], !empty($p['nested']) ? 1 : 0]; }, $params);
        $canon = json_encode(['auth' => $authed ? 1 : 0, 'method' => strtoupper($method), 'params' => $canonParams, 'route' => $route], JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
        return ['route' => $route, 'method' => strtoupper($method), 'authenticated' => $authed, 'status' => $status, 'params' => $params, 'shape' => self::fnv1a($canon)];
    }

    private static function send(string $token, array $batch): ?array
    {
        if (!function_exists('curl_init')) return null;
        $ch = curl_init(self::endpoint());
        curl_setopt_array($ch, [
            CURLOPT_POST => true,
            CURLOPT_HTTPHEADER => ['Authorization: Bearer ' . $token, 'Content-Type: application/json'],
            CURLOPT_POSTFIELDS => json_encode(['sketches' => $batch]),
            CURLOPT_RETURNTRANSFER => true, CURLOPT_TIMEOUT => 3, CURLOPT_CONNECTTIMEOUT => 2,
        ]);
        $res = @curl_exec($ch);
        curl_close($ch);
        return $res ? (json_decode($res, true) ?: null) : null;
    }

    private static function savePolicy(string $token, array $res): array
    {
        $p = [
            'mode' => $res['mode'] ?? 'observe',
            'shapes' => $res['policy']['shapes'] ?? [],
            'knownBad' => $res['policy']['knownBad'] ?? [],
        ];
        @file_put_contents(self::cacheFile($token), json_encode($p));
        return $p;
    }

    /** Load the cached policy, refreshing from the server if the cache is missing or stale. */
    private static function policy(string $token): array
    {
        $f = self::cacheFile($token);
        if (is_file($f) && (time() - filemtime($f)) < self::TTL) {
            $d = json_decode(file_get_contents($f), true);
            if (is_array($d)) return $d;
        }
        $res = self::send($token, []); // refresh
        return $res ? self::savePolicy($token, $res) : (['mode' => 'observe', 'shapes' => [], 'knownBad' => []]);
    }

    private static function decide(array $policy, string $shape): array
    {
        $per = $policy['shapes'][$shape] ?? null;
        if ($per === 'allow') return [false, null];
        if ($per === 'block') return [true, 'policy: blocked shape'];
        if (in_array($shape, $policy['knownBad'] ?? [], true)) return [true, 'global threat intelligence'];
        if (!empty($policy['shapes'])) return [true, 'off-baseline: unapproved behavior'];
        return [false, null];
    }

    /**
     * Non-echoing decision for framework middleware (Laravel/Symfony): returns [block(bool),
     * reason(?string)]. Records the blocked attempt when it blocks.
     */
    public static function verdict(string $token, string $method, string $path, array $query, bool $authed): array
    {
        if (self::neverBlock($path)) return [false, null]; // break-glass: never block the auth path
        $policy = self::policy($token);
        if (($policy['mode'] ?? 'observe') === 'enforce') {
            $sk = self::buildSketch($method, $path, $query, $authed, 0);
            [$block, $reason] = self::decide($policy, $sk['shape']);
            if ($block) {
                self::send($token, [self::buildSketch($method, $path, $query, $authed, 403)]);
                return [true, $reason];
            }
        }
        return [false, null];
    }

    /** Call at the START of a request (raw PHP). Returns true (and emits a 403) if blocked. */
    public static function guard(string $token, ?string $method = null, ?string $path = null, ?array $query = null, ?bool $authed = null): bool
    {
        $method = $method ?? ($_SERVER['REQUEST_METHOD'] ?? 'GET');
        $path = $path ?? strtok($_SERVER['REQUEST_URI'] ?? '/', '?');
        $query = $query ?? ($_GET ?? []);
        $authed = $authed ?? (isset($_SERVER['HTTP_AUTHORIZATION']) || isset($_SESSION['user']) || isset($_COOKIE['session']));
        [$block, $reason] = self::verdict($token, $method, $path, $query, $authed);
        if ($block) {
            http_response_code(403);
            header('content-type: application/json');
            echo json_encode(['error' => 'blocked_by_nemesis_shield', 'reason' => $reason]);
            return true;
        }
        return false;
    }

    /** Call at the END of a request to record the observed behavior. */
    public static function observe(string $token, ?string $method = null, ?string $path = null, ?array $query = null, ?bool $authed = null, ?int $status = null): void
    {
        $method = $method ?? ($_SERVER['REQUEST_METHOD'] ?? 'GET');
        $path = $path ?? strtok($_SERVER['REQUEST_URI'] ?? '/', '?');
        $query = $query ?? ($_GET ?? []);
        $authed = $authed ?? (isset($_SERVER['HTTP_AUTHORIZATION']) || isset($_SESSION['user']) || isset($_COOKIE['session']));
        $status = $status ?? (http_response_code() ?: 200);
        $res = self::send($token, [self::buildSketch($method, $path, $query, $authed, $status)]);
        if ($res) self::savePolicy($token, $res);
    }
}

File: ruby/nemesis_shield.rb
# Nemesis Shield - Sentinel SDK for Ruby (native: local shape + policy cache + inline blocking).
# Learns your app's normal behavior; in enforce mode blocks off-baseline requests (auth bypass, path
# traversal, scanners, unusual methods) before your app runs. Works with any Rack app - Rails,
# Sinatra, Hanami, raw Rack. Fail-open, privacy-preserving.
#
#   # Rails (config/application.rb):     config.middleware.use NemesisShield::Middleware, token: ENV["NEMESIS_TOKEN"]
#   # Sinatra:                           use NemesisShield::Middleware, token: ENV["NEMESIS_TOKEN"]
require "net/http"
require "json"
require "uri"
require "thread"

module NemesisShield
  DEFAULT_ENDPOINT = "https://shield.nemesislabs.xyz/api/v1/sketches".freeze
  # Canonical value taxonomy - must match the shared engine (tokenize.ts) byte-for-byte so a Ruby
  # app and a Node/Python app produce identical shape hashes.
  UUID = /\A[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\z/i
  EMAIL = /\A[^@\s]+@[^@\s]+\.[^@\s]+\z/
  URLRE = /\Ahttps?:\/\/\S+\z/i
  IPV4 = /\A(\d{1,3}\.){3}\d{1,3}\z/
  IPV6 = /\A[0-9a-f:]+:[0-9a-f:]+\z/i
  INT = /\A-?\d+\z/
  FLOATRE = /\A-?\d*\.\d+\z/
  DATE = /\A\d{4}-\d{2}-\d{2}([T ]\d{2}:\d{2}(:\d{2})?)?/
  JWT = /\A[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\z/
  HEX = /\A[0-9a-f]+\z/i
  B64 = /\A[A-Za-z0-9+\/]+={0,2}\z/
  ALPHA = /\A[A-Za-z]+\z/
  ALNUM = /\A[A-Za-z0-9]+\z/
  # A run of 7+ letters => reads like a word / route name, not an opaque id/token.
  WORD = /[A-Za-z]{7,}/
  # A hostname path segment (network-zone routes like example.com, sub.example.co.uk) collapses to
  # {domain}. Lookahead-free so it ports byte-identically to RE2 engines and matches the edge SDK.
  DOMAIN = /\A([a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?\.)+[a-z]{2,24}\z/i
  # A digit, used to tell a generated composite id (inc_ip_1_2_3_4_178..) from a snake_case word.
  DIGIT = /\d/

  module_function

  def normalize_path(path)
    clean = path.to_s.split("?", 2).first.to_s.split("#", 2).first.to_s
    out = clean.split("/").map do |seg|
      next seg if seg.empty?
      next "{traversal}" if seg.include?("..") # path-traversal segment (keeps ".." out of telemetry)
      case kind_of(seg)
      when "int", "float" then "{int}"
      when "uuid" then "{uuid}"
      when "hex" then "{hex}"
      when "base64" then "{token}"
      when "alnum" then (seg.length >= 12 && seg !~ WORD) ? "{id}" : seg
      else
        # Hostnames collapse to {domain}; underscored generated ids that carry a digit (or are very
        # long) collapse to {id}. Kebab route names (iso-27001) have no underscore, so stay literal.
        if seg =~ DOMAIN then "{domain}"
        elsif seg.include?("_") && (seg =~ DIGIT || seg.length >= 20) then "{id}"
        else seg end
      end
    end.join("/")
    out.empty? ? "/" : out
  end

  def kind_of(v)
    return "bool" if v == true || v == false
    return "empty" if v.nil?
    s = v.to_s
    return "empty" if s.empty?
    return "uuid" if s =~ UUID
    return "email" if s =~ EMAIL
    return "url" if s =~ URLRE
    return "ipv4" if s =~ IPV4
    return "ipv6" if s =~ IPV6
    return "jwt" if s =~ JWT
    return "date" if s =~ DATE
    return "int" if s =~ INT
    return "float" if s =~ FLOATRE
    return "hex" if s.length >= 16 && s =~ HEX
    return "base64" if s.length >= 16 && s =~ B64 && s !~ WORD
    return "alpha" if s =~ ALPHA
    return "alnum" if s =~ ALNUM
    "string"
  end

  def fnv1a(str)
    h = 0x811c9dc5
    str.each_byte { |b| h ^= b; h = (h * 0x01000193) & 0xffffffff }
    format("%08x", h)
  end

  # Request signature (method + route + query param kinds + auth). Excludes status by design.
  # Analytics / click-tracking query params carry no application logic and are the main cause of shape
  # explosion (every ad/campaign link adds different ones). Stripped from the signature so a UTM'd
  # request matches the bare route, while real params (and any attack in them) stay modeled. Shared,
  # identical across every Nemesis Shield SDK so the shape hash matches everywhere.
  TRACKING_PREFIXES = %w[utm_ mtm_ pk_ hsa_ matomo_ piwik_ ga_].freeze
  TRACKING_EXACT = %w[gclid gbraid wbraid dclid gclsrc fbclid msclkid twclid ttclid yclid igshid scid wickedid _ga _gl _hsenc _hsmi mc_cid mc_eid vero_id vero_conv oly_anon_id oly_enc_id _openstat rb_clickid s_cid epik sccid].freeze

  def tracking_param?(name)
    n = name.to_s.downcase
    TRACKING_PREFIXES.any? { |p| n.start_with?(p) } || TRACKING_EXACT.include?(n)
  end
  module_function :tracking_param?

  def build_sketch(method:, path:, query: {}, authed: false, status: 0)
    route = normalize_path(path)
    params = (query || {}).keys.reject { |k| tracking_param?(k) }.sort.map { |k| v = query[k]; { name: k.to_s, kind: kind_of(v.is_a?(Array) ? v.first : v), nested: v.is_a?(Array) } }
    # canonical shape input: keys sorted (auth, method, params, route); params are [name, kind, nested];
    # status is intentionally excluded so enforcement can decide BEFORE the response exists.
    canon = JSON.generate({ auth: authed ? 1 : 0, method: method.to_s.upcase, params: params.map { |p| [p[:name], p[:kind], p[:nested] ? 1 : 0] }, route: route })
    { route: route, method: method.to_s.upcase, authenticated: authed, status: status, params: params, shape: fnv1a(canon) }
  end

  # Safe-unlock (break-glass): paths never blocked, so a still-learning baseline can't lock operators
  # out of the doors they need to fix it. Prefix-matched. Override with the NEMESIS_SHIELD_BOOTSTRAP
  # env (comma-separated) or the `bootstrap:` option.
  DEFAULT_BOOTSTRAP = ["/login", "/signin", "/sign-in", "/auth", "/oauth", "/session", "/wp-login.php", "/wp-admin"].freeze

  def resolve_bootstrap(cfg)
    return cfg.map(&:to_s) if cfg.is_a?(Array)
    env = ENV["NEMESIS_SHIELD_BOOTSTRAP"].to_s
    return env.split(",").map(&:strip).reject(&:empty?) unless env.strip.empty?
    DEFAULT_BOOTSTRAP.dup
  end
  module_function :resolve_bootstrap

  class Client
    def initialize(token, endpoint: DEFAULT_ENDPOINT, flush_interval: 2, bootstrap: nil)
      @token = token
      @uri = URI(endpoint)
      @mode = "observe"
      @shapes = {}
      @known_bad = []
      @baseline = false
      @buffer = []
      @bootstrap = NemesisShield.resolve_bootstrap(bootstrap)
      @mu = Mutex.new
      if token && flush_interval && flush_interval > 0
        refresh
        Thread.new do
          loop { sleep flush_interval; flush; refresh }
        end
      end
    end

    def enforcing?
      @mu.synchronize { @mode == "enforce" }
    end

    # Break-glass: is this path on the never-block bootstrap allow-list? Prefix match.
    def never_block(path)
      p = path.to_s.split("?").first.to_s.split("#").first.to_s.downcase
      @bootstrap.any? { |b| !b.empty? && p.start_with?(b.downcase) }
    end

    def decide(sketch)
      @mu.synchronize do
        per = @shapes[sketch[:shape]]
        return [false, nil] if per == "allow"
        return [true, "policy: blocked shape"] if per == "block"
        return [true, "global threat intelligence"] if @known_bad.include?(sketch[:shape])
        return [true, "off-baseline: unapproved behavior"] if @baseline
        [false, nil]
      end
    end

    def record(sketch)
      flush_now = false
      @mu.synchronize { @buffer << sketch; flush_now = @buffer.size >= 50 }
      flush if flush_now
    end

    def flush
      batch = @mu.synchronize { b = @buffer; @buffer = []; b }
      send_batch(batch) unless batch.empty?
    end

    def refresh
      send_batch([])
    end

    def send_batch(batch)
      http = Net::HTTP.new(@uri.host, @uri.port)
      http.use_ssl = @uri.scheme == "https"
      http.open_timeout = 2
      http.read_timeout = 3
      req = Net::HTTP::Post.new(@uri, "Authorization" => "Bearer #{@token}", "Content-Type" => "application/json")
      req.body = JSON.generate({ sketches: batch })
      res = http.request(req)
      return unless res.is_a?(Net::HTTPSuccess)
      data = JSON.parse(res.body)
      @mu.synchronize do
        @mode = data["mode"] if data["mode"]
        pol = data["policy"] || {}
        if pol["shapes"]
          @shapes = pol["shapes"]
          @baseline = true unless @shapes.empty?
        end
        @known_bad = pol["knownBad"] if pol["knownBad"]
      end
    rescue StandardError
      nil # fail-open
    end
  end

  # Rack middleware (Rails / Sinatra / any Rack app).
  class Middleware
    def initialize(app, token:, endpoint: DEFAULT_ENDPOINT, **kw)
      @app = app
      @client = Client.new(token, endpoint: endpoint, **kw)
    end

    def call(env)
      method = env["REQUEST_METHOD"]
      path = env["PATH_INFO"].to_s
      authed = !env["HTTP_AUTHORIZATION"].nil? || !env["HTTP_COOKIE"].nil? || !env["HTTP_X_API_KEY"].nil?
      query = begin
        require "rack/utils"
        Rack::Utils.parse_query(env["QUERY_STRING"].to_s)
      rescue StandardError
        # stdlib fallback so query-param STRUCTURE is still fed when rack isn't loaded
        begin
          URI.decode_www_form(env["QUERY_STRING"].to_s).to_h
        rescue StandardError
          {}
        end
      end
      if @client.enforcing? && !@client.never_block(path)
        block, reason = @client.decide(NemesisShield.build_sketch(method: method, path: path, query: query, authed: authed))
        if block
          @client.record(NemesisShield.build_sketch(method: method, path: path, query: query, authed: authed, status: 403))
          body = JSON.generate({ error: "blocked_by_nemesis_shield", reason: reason })
          return [403, { "content-type" => "application/json" }, [body]]
        end
      end
      status, headers, response = @app.call(env)
      @client.record(NemesisShield.build_sketch(method: method, path: path, query: query, authed: authed, status: status))
      [status, headers, response]
    end
  end
end
